File: apps/api/app/agents/citation_whitelist.py
"""Curated whitelist of canonical Indian indirect/direct tax authorities.

Recurring partner feedback (Sprint 5 review on three test drafts):
    "The verifier has a recurring false negative on Pushpam Pharmaceuticals
     across two drafts. The case is real, foundational, and routinely cited
     in s.74 GST adjudications. Same risk exists for Cosmic Dye Chemical,
     Continental Foundation, Anand Nishikawa, Suncraft Energy, D.Y. Beathel
     Enterprises, On Quest Merchandising and other parallel authorities on
     the same point."

So instead of asking IndianKanoon's free-tier search to consistently surface
these (it doesn't — the corpus has hundreds of "Pushpam"-name hits and the
right one ranks below noise), we maintain a small curated map of widely-
cited tax-law authorities keyed by case-name substring. A match here always
returns VERIFIED with the canonical IndianKanoon URL, bypassing the network
call entirely. The verifier consults this list first; if no match, it falls
back to the configured provider (stub or IndianKanoon search).

This is NOT a substitute for real verification of obscure / new cases —
those still go to IndianKanoon. It's a pragmatic safety net for the dozen
or so cases every Indian tax practitioner uses every day.

The list is hand-curated; add new entries as partners flag false negatives.
Keys must be lowercase substrings that uniquely identify the case in the
universe of likely citation text.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class CanonicalCase:
    name: str
    canonical_url: str
    proposition: str  # short note on what the case stands for


# Substring (lowercased) → CanonicalCase. Substring match because the model
# may format the case name with or without years / reporter abbreviations.
CANONICAL_CASES: dict[str, CanonicalCase] = {
    # ---- Excise / GST suppression + extended period -----------------------
    "pushpam pharmaceuticals": CanonicalCase(
        name="Pushpam Pharmaceuticals Co. v. Collector of Central Excise, Bombay",
        canonical_url="https://indiankanoon.org/doc/1690509/",
        proposition=(
            "Suppression of facts under the proviso to s.11A Central Excise "
            "(now mirrored in s.74 CGST) requires deliberate withholding; "
            "mere non-disclosure is insufficient."
        ),
    ),
    "cosmic dye chemical": CanonicalCase(
        name="Cosmic Dye Chemical v. Collector of Central Excise",
        canonical_url="https://indiankanoon.org/doc/1623537/",
        proposition=(
            "Extended period of limitation cannot be invoked without intent "
            "to evade tax; mens rea is essential."
        ),
    ),
    "continental foundation": CanonicalCase(
        name="Continental Foundation Joint Venture v. CCE Chandigarh",
        canonical_url="https://indiankanoon.org/doc/1107015/",
        proposition=(
            "Suppression must be conscious and deliberate; a bona fide "
            "interpretation that is later disagreed with is not suppression."
        ),
    ),
    "anand nishikawa": CanonicalCase(
        name="Anand Nishikawa Co. Ltd. v. Commissioner of Central Excise",
        canonical_url="https://indiankanoon.org/doc/1257203/",
        proposition=(
            "Suppression of facts in the proviso to s.11A is to be construed "
            "strictly; the department must show wilful misstatement."
        ),
    ),
    # ---- Cross-examination / natural justice ------------------------------
    "andaman timber": CanonicalCase(
        name="Andaman Timber Industries v. Commissioner of Central Excise",
        canonical_url="https://indiankanoon.org/doc/24400737/",
        proposition=(
            "Denial of cross-examination of witnesses whose statements are "
            "relied upon vitiates the adjudication order."
        ),
    ),
    # ---- ITC mismatch / supplier default ---------------------------------
    "suncraft energy": CanonicalCase(
        name="Suncraft Energy Pvt Ltd v. Assistant Commissioner, State Tax",
        canonical_url="https://indiankanoon.org/doc/108089236/",
        proposition=(
            "ITC cannot be denied to the recipient merely because the "
            "supplier failed to deposit tax; recovery must first be "
            "attempted from the defaulting supplier (Calcutta HC, 2023)."
        ),
    ),
    "d.y. beathel": CanonicalCase(
        name="D.Y. Beathel Enterprises v. State Tax Officer",
        canonical_url="https://indiankanoon.org/doc/108580056/",
        proposition=(
            "Where supplier has not paid tax, reversal of ITC against the "
            "buyer without investigation of the supplier is unsustainable "
            "(Madras HC, 2021)."
        ),
    ),
    "on quest merchandising": CanonicalCase(
        name="On Quest Merchandising India Pvt Ltd v. Government of NCT of Delhi",
        canonical_url="https://indiankanoon.org/doc/115886907/",
        proposition=(
            "Section 9(2)(g) DVAT (analogous to GST ITC matching rules) "
            "cannot deny ITC to a purchaser for the seller's default "
            "(Delhi HC)."
        ),
    ),
    # ---- ITC entitlement principle ---------------------------------------
    "bharti airtel": CanonicalCase(
        name="Commissioner of GST v. Bharti Airtel Ltd",
        canonical_url="https://indiankanoon.org/doc/170611621/",
        proposition=(
            "The Department bears the burden of establishing a demand on "
            "positive material; mere book entries or system mismatches do "
            "not discharge that burden (SC, 2021)."
        ),
    ),
    "eicher motors": CanonicalCase(
        name="Eicher Motors Ltd v. Union of India",
        canonical_url="https://indiankanoon.org/doc/1486002/",
        proposition=(
            "Credit accrued is a vested right; subsequent legislation "
            "cannot retrospectively defeat it."
        ),
    ),
    # ---- Reconciliation defence ------------------------------------------
    "larsen": CanonicalCase(
        name="Larsen & Toubro Ltd v. State of Karnataka",
        canonical_url="https://indiankanoon.org/doc/89018960/",
        proposition=(
            "Reconciliation between returns is a sufficient response where "
            "records bear out the assessee's position (SC, 2014)."
        ),
    ),
    # ---- Show-cause notice formation -------------------------------------
    "amrit foods": CanonicalCase(
        name="Amrit Foods v. Commissioner of Central Excise, UP",
        canonical_url="https://indiankanoon.org/doc/1379149/",
        proposition=(
            "A show cause notice that does not specifically allege which "
            "limb of the proviso (fraud / collusion / wilful misstatement "
            "/ suppression) is invoked is bad in law."
        ),
    ),
    "uniworth textiles": CanonicalCase(
        name="Uniworth Textiles Ltd v. CCE, Raipur",
        canonical_url="https://indiankanoon.org/doc/164527889/",
        proposition=(
            "Mere non-payment of tax is not equivalent to suppression; "
            "extended period requires positive act of suppression."
        ),
    ),
    # ---- Direct tax: 148 reassessment ------------------------------------
    "ganga saran": CanonicalCase(
        name="ITO v. Ganga Saran & Sons (P) Ltd",
        canonical_url="https://indiankanoon.org/doc/1521085/",
        proposition=(
            "The 'reason to believe' for s.148 reopening must have a "
            "rational connection with the material on record; suspicion is "
            "not enough (SC)."
        ),
    ),
    "kelvinator": CanonicalCase(
        name="CIT v. Kelvinator of India Ltd",
        canonical_url="https://indiankanoon.org/doc/1192802/",
        proposition=(
            "Reassessment under s.148 cannot be based on a mere change of "
            "opinion; tangible material is required (SC, 2010)."
        ),
    ),
    "ashish agarwal": CanonicalCase(
        name="Union of India v. Ashish Agarwal",
        canonical_url="https://indiankanoon.org/doc/183568974/",
        proposition=(
            "Section 148A(b) procedure must be followed for reassessment "
            "notices issued under the new regime (SC, 2022)."
        ),
    ),
}


def lookup_canonical(case_name: str) -> CanonicalCase | None:
    """Substring-match a citation's case name against the whitelist."""
    needle = case_name.lower()
    for key, entry in CANONICAL_CASES.items():
        if key in needle:
            return entry
    return None

File: noticedesk-java/src/main/java/com/noticedesk/api/agent/CitationWhitelist.java
package com.noticedesk.api.agent;

import java.util.Map;

public final class CitationWhitelist {

    public record CanonicalCase(String name, String canonicalUrl, String proposition) {}

    private static final Map<String, CanonicalCase> CANONICAL_CASES = Map.ofEntries(
        Map.entry("pushpam pharmaceuticals", new CanonicalCase(
            "Pushpam Pharmaceuticals Co. v. Collector of Central Excise, Bombay",
            "https://indiankanoon.org/doc/1690509/",
            "Suppression of facts under the proviso to s.11A Central Excise (now mirrored in s.74 CGST) requires deliberate withholding; mere non-disclosure is insufficient."
        )),
        Map.entry("cosmic dye chemical", new CanonicalCase(
            "Cosmic Dye Chemical v. Collector of Central Excise",
            "https://indiankanoon.org/doc/1623537/",
            "Extended period of limitation cannot be invoked without intent to evade tax; mens rea is essential."
        )),
        Map.entry("continental foundation", new CanonicalCase(
            "Continental Foundation Joint Venture v. CCE Chandigarh",
            "https://indiankanoon.org/doc/1107015/",
            "Suppression must be conscious and deliberate; a bona fide interpretation that is later disagreed with is not suppression."
        )),
        Map.entry("anand nishikawa", new CanonicalCase(
            "Anand Nishikawa Co. Ltd. v. Commissioner of Central Excise",
            "https://indiankanoon.org/doc/1257203/",
            "Suppression of facts in the proviso to s.11A is to be construed strictly; the department must show wilful misstatement."
        )),
        Map.entry("andaman timber", new CanonicalCase(
            "Andaman Timber Industries v. Commissioner of Central Excise",
            "https://indiankanoon.org/doc/24400737/",
            "Denial of cross-examination of witnesses whose statements are relied upon vitiates the adjudication order."
        )),
        Map.entry("suncraft energy", new CanonicalCase(
            "Suncraft Energy Pvt Ltd v. Assistant Commissioner, State Tax",
            "https://indiankanoon.org/doc/108089236/",
            "ITC cannot be denied to the recipient merely because the supplier failed to deposit tax."
        )),
        Map.entry("d.y. beathel", new CanonicalCase(
            "D.Y. Beathel Enterprises v. State Tax Officer",
            "https://indiankanoon.org/doc/108580056/",
            "Where supplier has not paid tax, reversal of ITC against the buyer without investigation of the supplier is unsustainable."
        )),
        Map.entry("on quest merchandising", new CanonicalCase(
            "On Quest Merchandising India Pvt Ltd v. Government of NCT of Delhi",
            "https://indiankanoon.org/doc/115886907/",
            "Section 9(2)(g) DVAT cannot deny ITC to a purchaser for the seller's default."
        )),
        Map.entry("bharti airtel", new CanonicalCase(
            "Commissioner of GST v. Bharti Airtel Ltd",
            "https://indiankanoon.org/doc/170611621/",
            "The Department bears the burden of establishing a demand on positive material; mere book entries or system mismatches do not discharge that burden."
        )),
        Map.entry("eicher motors", new CanonicalCase(
            "Eicher Motors Ltd v. Union of India",
            "https://indiankanoon.org/doc/1486002/",
            "Credit accrued is a vested right; subsequent legislation cannot retrospectively defeat it."
        )),
        Map.entry("larsen", new CanonicalCase(
            "Larsen & Toubro Ltd v. State of Karnataka",
            "https://indiankanoon.org/doc/89018960/",
            "Reconciliation between returns is a sufficient response where records bear out the assessee's position."
        )),
        Map.entry("amrit foods", new CanonicalCase(
            "Amrit Foods v. Commissioner of Central Excise, UP",
            "https://indiankanoon.org/doc/1379149/",
            "A show cause notice that does not specifically allege which limb of the proviso is invoked is bad in law."
        )),
        Map.entry("uniworth textiles", new CanonicalCase(
            "Uniworth Textiles Ltd v. CCE, Raipur",
            "https://indiankanoon.org/doc/164527889/",
            "Mere non-payment of tax is not equivalent to suppression; extended period requires positive act of suppression."
        )),
        Map.entry("ganga saran", new CanonicalCase(
            "ITO v. Ganga Saran & Sons (P) Ltd",
            "https://indiankanoon.org/doc/1521085/",
            "The 'reason to believe' for s.148 reopening must have a rational connection with the material on record."
        )),
        Map.entry("kelvinator", new CanonicalCase(
            "CIT v. Kelvinator of India Ltd",
            "https://indiankanoon.org/doc/1192802/",
            "Reassessment under s.148 cannot be based on a mere change of opinion; tangible material is required."
        )),
        Map.entry("ashish agarwal", new CanonicalCase(
            "Union of India v. Ashish Agarwal",
            "https://indiankanoon.org/doc/183568974/",
            "Section 148A(b) procedure must be followed for reassessment notices issued under the new regime."
        ))
    );

    private CitationWhitelist() {}

    public static CanonicalCase lookupCanonical(String caseName) {
        String needle = caseName.toLowerCase();
        for (Map.Entry<String, CanonicalCase> entry : CANONICAL_CASES.entrySet()) {
            if (needle.contains(entry.getKey())) {
                return entry.getValue();
            }
        }
        return null;
    }
}
